Rework annotation ordering/optimisations

tecosaur opened the pull request:
There are some annoyances/discrepancies with annotation handling that I originally tried to fix in #53794, #53800 and #53801.

@LilithHafner was kind enough to have a look at these, and after a long chat we came to the conclusion that this isn't the right approach to take, and that the order with which annotations are applied should be given primacy over annotation ranges.

This PR supersedes the aforementioned PRs, and should make the way annotations are ordered more sensible.

See the commit messages for more details.

tecosaur committed:
Remove strong ordering of annotation ranges

After a long chat with Lilith Halfner, we've come to the conclusion that
the range-based ordering applied to annotations, while nice for making
some otherwise O(n) code O(log n) and O(n^2) code O(n), is actually
assuming too much about how annotations are used and interact with each
other. Removing all assumptions about ordering, and giving annotation
order primacy seems like the most sensible thing to do, even if it makes
a few bits of the code less algorithmically "nice".

As a consequence, we also get rid of annotatedstring_optimize!. Specific
producers/consumers of annotated text will know what assumptions can be
made to compress/optimise the annotations used, and are thus best suited
to do so themselves. The one exception to this is probably when writing
to an AnnotatedIOBuffer, here adding a specific optimisation to
_insert_annotations! probably makes sense, and will be explored soon.

tecosaur committed:
Make _insert_annotations! optimising

This is a rather important optimisation, since it prevents the
annotation blow-up that can result from say writing to an
AnnotatedIOBuffer char-by-char. Originally I was just going to pass the
AnnotatedString produced when reading the AnnotatedIOBuffer through
annotatedstring_optimize!, but now that's been removed, this seems like
the best past forwards (it's also actually a better approach than
applying annotatedstring_optimize!, just hard to justify when that code
already existed).

tecosaur committed:
Bump StyledStrings

This is in preparation for changes to the way annotation ordering is
handled in Base.

tecosaur commented:
Ah, tests are failing because I missed a use of `Base.annotatedstring_optimize!` in StyledStrings' tests. I'll fix that and re-bump the stdlib shortly.

bvdmitri commented on base/strings/annotated.jl:
```diff
@@ -19,7 +19,7 @@ annotated with labeled values.
 
 The above diagram represents a `AnnotatedString` where three ranges have been
-annotated (labeled `A`, `B`, and `C`). Each annotation holds a label (`Symbol`)
+annotated (labelled `A`, `B`, and `C`). Each annotation holds a label (`Symbol`)
```
These are not really typos but US vs UK way of writing.

tecosaur replied:
Ah, not this again... Thanks for letting me know, I'll just drop this commit.

tecosaur commented:
Ah yep, we did also discuss a need for more info in the docstrings. I'll see about adding some of the elaboration needed.

tecosaur committed:
Document the semantics of annotation ordering

It's important to specify the way that annotations relate to the
characters of the underlying string and each other.

Along the way, it's also worth explaining the behaviour of the internal
functions _clear_annotations_in_region! and _insert_annotations!.

LilithHafner reviewed (commented):
```julia
    for (region, annot) in str.annotations
        if region == fullregion
            push!(annotations, (firstindex(unannot):lastindex(unannot), annot))
        end
    end
    for offset in 0:len:(r-1)*len
        for (region, annot) in str.annotations
            if region != fullregion
                push!(annotations, (region .+ offset, annot))
            end
        end
    end
```
This is buggy in the new semantic because it re-orders annotations. Not necessarily blocking this PR, but an issue.

LilithHafner commented on base/strings/annotated.jl:
```diff
@@ -516,31 +498,58 @@ function _clear_annotations_in_region!(annotations::Vector{Tuple{UnitRange{Int},
+    end
+    # Insert any extra entries in the appropriate position
+    for (offset, (i, entry)) in enumerate(extras)
+        insert!(annotations, i + offset, entry)
```
This is O(n^2), but we can revise this function for perf later.

tecosaur replied:
Yep, and I don't think this should be in the "hot path" of any common use-cases :crossed_fingers:.

LilithHafner commented on base/strings/annotated.jl:
```diff
@@ -505,7 +487,7 @@ end
 function _clear_annotations_in_region!(annotations::Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}}, span::UnitRange{Int})
```
```suggestion
"""
    _clear_annotations_in_region!(annotations::Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}}, span::UnitRange{Int})

remove all annotations that fall within `span` by deleting ranges that are entirely contained, truncating
ranges that partially overlap, and splitting ranges that subsume span
"""
function _clear_annotations_in_region!(annotations::Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}}, span::UnitRange{Int})
```
Documenting internals makes it easier for me to review. If this description is right, please accept the suggestion. Otherwise, I'll need to reevaluate my understanding of the function.

tecosaur replied:
Mmmm, these functions _should_ all have docstrings. I'll see about adding a commit for this function and others.

tecosaur replied:
Docstrings added in `Document the semantics of annotation ordering`

LilithHafner commented on base/strings/annotated.jl:
```diff
@@ -516,31 +498,58 @@ function _clear_annotations_in_region!(annotations::Vector{Tuple{UnitRange{Int},
+            first(first(annot)) == 1 || continue
+            if last(annot) == last(last(io.annotations))
+                valid_run = true
+                for runlen in 1:i
```
This is also O(n^2) when it could be O(n), but perf can wait.

LilithHafner replied:
And it will be O(n) in non-pathological cases.

tecosaur replied:
Yea, this is written with the non-pathological case in mind, and if you consider the case where few-annotation insertions are repeatedly made, individual insertions should be pretty much `O(1)` with a constant factor based on the average "checked annotation depth" / average number of annotations each insertion.

LilithHafner commented on base/strings/annotated.jl:
```diff
@@ -516,31 +498,58 @@ function _clear_annotations_in_region!(annotations::Vector{Tuple{UnitRange{Int},
     annotations
 end
 
 function _insert_annotations!(io::AnnotatedIOBuffer, annotations::Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}}, offset::Int = position(io))
```
Does this function have unit tests?

tecosaur replied:
Yes, it's checked by the new block of tests at the end of `test/strings/annotated.jl`, under ``# The `_insert_annotations!` cautious-merging optimisation``.

LilithHafner reviewed (approved):
I think this should be merged and followup prs should address the performance of this PR's implementations and address the now-invalid-ness of the repeat function.

tecosaur commented:
I've just adjusted `repeat` to be consistent with the new semantic,
```julia
fullregion = firstindex(str):lastindex(str)
if allequal(first, str.annotations) && first(first(str.annotations)) == fullregion
    newfullregion = firstindex(unannot):lastindex(unannot)
    for (_, annot) in str.annotations
        push!(annotations, (newfullregion, annot))
    end
else
    for offset in 0:len:(r-1)*len
        for (region, annot) in str.annotations
            push!(annotations, (region .+ offset, annot))
        end
    end
end
```
this change has been folded into `2656884453 * Remove strong ordering of annotation ranges`

tecosaur commented:
Now that #54308 has been merged, we need to resolve the merge conflict and adjust the test introduced there.

tecosaur commented:
After CI, I think this should be good to merge if that also sounds good by @LilithHafner :slightly_smiling_face:.

LilithHafner commented:
1bed84f4ae6fe6fc9bb39612e274fd1630a51349 LGTM. I have not reviewed https://github.com/JuliaLang/julia/compare/1bed84f4ae6fe6fc9bb39612e274fd1630a51349..4c18472b3e56998f2784ecd5d9deea4cf5371850, but am open to suggestions to how I could do that without re-reviewing this PR from scratch.

LilithHafner commented:
Okay, the diff in your comment also LGTM.

IanButterworth commented:
@tecosaur I know you have a preference to not squash but for things that will get backported, it introduces unnecessary stumbling blocks in that automated process that would be better to avoid.

tecosaur commented:
Thanks for mentioning that Ian, I do like to preserve more granular information when it seems sensible, but I wasn't aware that complicated the backporting process. I'll keep that in mind with the few remaining PRs I'm hoping to get backported :slightly_smiling_face:.

(As an aside, is the automated backporting doing something other than cherry picking?)

IanButterworth commented:
Yeah the automatic script doesn't handle adding `-m`. If you wanted to help with adding that it's here https://github.com/KristofferC/Backporter/blob/master/backporter.jl
